### PR TITLE
Try building ABI3 wheels for cpython

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: .ci/scripts/auditwheel_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        subprocess.run(f"delocate-listdeps {wheel_file}", shell=True, check=True)
+        subprocess.run(
+            f"delocate-wheel --require-archs {archs} -w {dest_dir} {wheel_file}",
+            shell=True,
+            check=True,
+        )
+    else:
+        subprocess.run(
+            f"auditwheel repair -w {dest_dir} {wheel_file}", shell=True, check=True
+        )
```
Comment (DMRobertson) [inline .ci/scripts/auditwheel_wrapper.py:104]: Can we do without `shell=True`---it's a bit of a security red flag IIRC?

Also where do `delocate-listdeps`, `delocate-wheel` and `auditwheel` come from---are they installed with `cibuildwheel`?

File: .ci/scripts/auditwheel_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        raise Exception(f"Unexpected build tag: {build}")
+
+    # If the wheel is for cpython then convert it into an abi3 wheel.
+    if tag.interpreter.startswith("cp"):
+        wheel_file = cpython(wheel_file, name, version, tag)
+
+    # Finally, repair the wheel.
+    if archs is not None:
+        # If we are given archs then we are on macos and need to use
+        # `delocate-listdeps`.
```
Comment (DMRobertson) [inline .ci/scripts/auditwheel_wrapper.py:94]: Is this true in the light of the recent aarch64 changes?

**Review by DMRobertson — CHANGES_REQUESTED** [review]
Do these wheels still get tested by cibuildwheel?

Are we reducing the total number of wheels we build here---I don' think we are? If not, what's the motivation here?

**Comment by erikjohnston** [comment]
> Do these wheels still get tested by cibuildwheel?
> 
> Are we reducing the total number of wheels we build here---I don' think we are? If not, what's the motivation here?

Yup, cibuildwheel simply skips a build step if it finds an existing ABI compatible wheel. It'll still do the test step.

See https://github.com/matrix-org/synapse/actions/runs/3296929336 for a full run

File: .ci/scripts/auditwheel_wrapper.py (hunk 1 above)
Comment (erikjohnston) [reply .ci/scripts/auditwheel_wrapper.py:104]: Yeah, they already exist in the containers. These commands are actually the default repair commands, so are run today.

File: .ci/scripts/auditwheel_wrapper.py (hunk 2 above)
Comment (erikjohnston) [reply .ci/scripts/auditwheel_wrapper.py:94]: Yes, see https://cibuildwheel.readthedocs.io/en/stable/options/#repair-wheel-command

**Comment by DMRobertson** [comment]
> Yup, cibuildwheel simply skips a build step if it finds an existing ABI compatible wheel.

Brilliant, thanks for confirming!

File: .ci/scripts/auditwheel_wrapper.py (hunk 1 above)
Comment (erikjohnston) [reply .ci/scripts/auditwheel_wrapper.py]: > Can we do without `shell=True`---it's a bit of a security red flag IIRC?

Done, though in this instance its not a security problem as the args are given to us via cibuildwheel, which already can call arbitrary shell commands.

https://docs.python.org/3/library/subprocess.html#security-considerations

**Review by DMRobertson — APPROVED** [review]
Sorry for pedantry. LGTM!

**Comment by clokep** [comment]
Follow-up typo fix: 1e73effebf4f307f499c8cd4be58f805959beb38

**Pull request merged**